### DISCUSSION TITLE
Equivalence of bash built-in command `trap` in Xonsh?

### DISCUSSION THREAD

**Discussion by nukemiko** [xonsh/xonsh#5039] [Q&A] [answered]
I need to do something in my Xonsh script while Xonsh exiting, such as unmount or delete something. In bash, it has a built-in command `trap` can do it, what is the equivalence of `trap` in Xonsh?

P.S. I tried the Python standard library module `atexit`, but if the function used for `atexit.register()` needs to run something through Subprocess-mode, it will raise an exception on exiting:

```xonsh
import atexit

@atexit.register
def cleanup():
    print('Running cleanup (Python-mode)...')
    echo 'Running cleanup (Subprocess-mode)...'
    seq 10
```

```
Running cleanup (Python-mode)...
Exception ignored in atexit callback: <function cleanup at 0x7f3dd61f7d00>
Traceback (most recent call last):
  File "<stdin>", line 4, in cleanup
NameError: name '__xonsh__' is not defined
```

**Comment by anki-code** [comment]
> Trap allows you to catch signals and execute code when they occur.

For the [docker-xonsh-wrapper](https://github.com/anki-code/docker-xonsh-wrapper) I wrote [the code that catch SIGINT and SIGTERM](https://github.com/anki-code/docker-xonsh-wrapper/blob/03a1e0372b18358c556c094dc8a6a29c43d3f126/run.xsh).

**Comment by gforsyth** [comment, marked as answer]
You can use the built-in events system for this:
```
@events.on_exit 
def _(): 
    print("running cleanup (python mode") 
    echo "running subproc cleanup" 
    seq 10 
```

see https://xon.sh/events.html#on-exit and https://xon.sh/tutorial_events.html
